### PR TITLE
[Improvement] Support soft label for CrossEntropyLoss

### PR DESCRIPTION
**Pull request by irvingzhang0512**
## Motivation
Soft labels are required for data augmentation methods like mixup, cutmix, etc. For now, official `CrossEntropyLoss`  doesn't support soft labels.

## Details
+ If number of classes is 4.
    + Hard labels example: `[0, 1, 2]`
    + Corresponding soft labels are `[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0]]`
    + Mixup/Cutmix may create soft labels like `[[0.8, 0.2, 0, 0], [0.1, 0.9, 0, 0], [0, 0, 0.6, 0.4]]`
+ It's difficult to get a proper name for this pr and related args.
    + `Sparse` is borrowed from TensorFlow and MXNet, TensorFlow has  `CategoricalCrossentropy`(accept non-sparse, one-hot like labels) and `SparseCategorialCrossentropy`(just like `CrossEntropyLoss` in pytorch).
    + BTW, **one-hot like labels** is more clear and meaningful for me.
+ For now, this pr detects loss type by the shape of label.

## TODO
+ [x] Modify `CrossEntropyLoss` in `mmaction/models/losses/cross_entropy_loss.py`.
+ [x] unittest
+ [x]  changelog.

### CODE REVIEW

File: mmaction/models/losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -16,13 +16,20 @@ class CrossEntropyLoss(BaseWeightedLoss):
             as None, use the same weight 1 for all classes. Only applies
             to CrossEntropyLoss and BCELossWithLogits (should not be set when
             using other losses). Default: None.
+        sparse_label (bool): Whether to use sparse labels or not. Sparse labels
```
Comment (innerlee) [inline mmaction/models/losses/cross_entropy_loss.py:19]: How about detecting which kind of loss from the shape of label?

Comment (irvingzhang0512) [reply mmaction/models/losses/cross_entropy_loss.py:19]: Sounds good. This could also simplify the codes of mixup/cutmix.

File: mmaction/models/losses/cross_entropy_loss.py [hunk 2]
```diff
@@ -9,6 +9,16 @@
 class CrossEntropyLoss(BaseWeightedLoss):
     """Cross Entropy Loss.
 
+    Support two kinds of labels and their corresponding loss type. It's worth
+    mentioning that loss type will be detected by the shape of ``cls_score``
+    and ``label``.
+    1) Sparse label: integers in range [0, num_classes - 1].
+    2) Non-sparse label(one-hot like labels): floats in range [0, 1]. It shares
```
Comment (innerlee) [inline mmaction/models/losses/cross_entropy_loss.py:16]: What's the difference between soft-label and non-sparse label?

Comment (irvingzhang0512) [reply mmaction/models/losses/cross_entropy_loss.py:16]: I couldn't tell the difference between non-sparse label/soft-label/one-hot like label. I just don't know which term is more former and meaningful.

Comment (innerlee) [reply mmaction/models/losses/cross_entropy_loss.py:16]: let's use soft-label

File: mmaction/models/losses/cross_entropy_loss.py [hunk 3]
```diff
@@ -9,6 +9,20 @@
 class CrossEntropyLoss(BaseWeightedLoss):
     """Cross Entropy Loss.
 
+    Support two kinds of labels and their corresponding loss type. It's worth
+    mentioning that loss type will be detected by the shape of ``cls_score``
+    and ``label``.
+    1) Sparse label: This label is an integer array and all of the elements are
```
Comment (innerlee) [inline mmaction/models/losses/cross_entropy_loss.py:15]: Sparse -> One-hot

File: mmaction/models/losses/cross_entropy_loss.py [hunk 4]
```diff
@@ -36,10 +50,33 @@ def _forward(self, cls_score, label, **kwargs):
         Returns:
             torch.Tensor: The returned CrossEntropy loss.
         """
-        if self.class_weight is not None:
-            assert 'weight' not in kwargs, "The key 'weight' already exists."
-            kwargs['weight'] = self.class_weight.to(cls_score.device)
-        loss_cls = F.cross_entropy(cls_score, label, **kwargs)
+        if cls_score.size() == label.size():
+            # cal loss for soft label
```
Comment (innerlee) [inline mmaction/models/losses/cross_entropy_loss.py:54]: One thing to take care is the `kwargs`. Please check that no kwargs are ignored silently

File: mmaction/models/losses/cross_entropy_loss.py (hunk 3 above)
Comment (irvingzhang0512) [reply mmaction/models/losses/cross_entropy_loss.py:15]: Non-sparse labels are one-hot like labels, aka a probability distribution.

Sparse label is a list of integers.

Comment (innerlee) [reply mmaction/models/losses/cross_entropy_loss.py:15]: I think this is a misuse of "sparse". How about just "labels"

Comment (irvingzhang0512) [reply mmaction/models/losses/cross_entropy_loss.py:15]: I'm not sure about the misuse thing.

[TF SparseCategoricalCrossentropy](https://www.tensorflow.org/api_docs/python/tf/keras/losses/SparseCategoricalCrossentropy)

[Book: Hands-On Machine Learning book with Scikit-Learn and TF](https://stats.stackexchange.com/questions/427748/what-does-it-mean-when-someone-says-they-have-sparse-labels)

[GluonCV](https://github.com/dmlc/gluon-cv/blob/2da042f8982b8c46cdf04b0646c845f1175402ef/gluoncv/loss.py#L275)

Comment (innerlee) [reply mmaction/models/losses/cross_entropy_loss.py:15]: It seems that the usage comes from tf. It is not good term, lets stop its popularization.

Okay joking :) I think "label" is clear enough here. If you want to differentiate, use "hard label"

File: mmaction/models/losses/cross_entropy_loss.py [hunk 5]
```diff
@@ -9,6 +9,17 @@
 class CrossEntropyLoss(BaseWeightedLoss):
     """Cross Entropy Loss.
 
+    Support two kinds of labels and their corresponding loss type. It's worth
+    mentioning that loss type will be detected by the shape of ``cls_score``
+    and ``label``.
+    1) Hard label: This label is an integer array and all of the elements are
+        in the range [0, num_classes - 1]. This label's shape should be
+        ``cls_score``'s shape with the `num_classes` dimension removed.
+    2) Soft label(one-hot like label): This label is a probability distribution
```
Comment (innerlee) [inline mmaction/models/losses/cross_entropy_loss.py:18]: ~~(one-hot like label)~~ (using probablity distribution over classes as label)

`One-hot like` gives more confusion than clarification. If something resembles one-hot, maybe it is "Two-hot" or "One-hotter/cooler"?
"The label shape is the similar to one-hot encoding" is the last thing one may think of

File: mmaction/models/losses/cross_entropy_loss.py (hunk 4 above)
Comment (innerlee) [reply mmaction/models/losses/cross_entropy_loss.py:51]: cal -> calculate

**Comment by innerlee** [comment]
Some minor comments.

**Review by innerlee — APPROVED** [review]

**Comment by kennymckormick** [comment]
Thanks for the PR~ Besides have u find some settings in which soft labels work in action recognition? I have tried to find one but failed. If you find such a setting, you can add the config file which uses soft label for training and achieves better results.

**Review by kennymckormick — APPROVED** [review]

**Comment by irvingzhang0512** [comment]
> Thanks for the PR~ Besides have u find some settings in which soft labels work in action recognition? I have tried to find one but failed. If you find such a setting, you can add the config file which uses soft label for training and achieves better results.

I plan to implement mixup/cutmix in two weeks, which use soft label.
image
